### PR TITLE
DEVX-1580 Provide some initial toolchain enhancements

### PR DESCRIPTION
**Pull request by bbejeck**
This PR provides an initial attempt to provide some utilities for tutorial authors.

Here's the TLDR;

`gen_project.sh` 

Generates the project scaffold with as much content as feasible.  To use
```
./gen_project.sh /some/directory/my_tutorial_properties.sh
```
Where `my_tutorial_properties.sh` would look like
```
#!/bin/sh

# space separated list of types, could just be one of ksql or kstreams
CARDS="kstreams ksql"

# all lower-case with no spaces
TUTORIAL_SHORT_NAME=my-tutorial-name

# also will become a test named FilteringCodeTest
MAIN_CLASS=FilteringCode
AK_VERSION=2.4.0
CP_VERSION=5.4.0
KSQLDB_VERSION=0.7.1
SEMAPHORE_TEST_NAME="Test name for Semaphore run"
PERMALINK="seo-friendly-link-to-my-tutorial"
```

`clone-tutorial.sh`
Clones an existing tutorial, changing the tutorial name in the content.  Great for new tutorial that similar to exiting ones.

```
./clone-tutorial.sh some/directory/clone_propteries.sh
```
Sample clone properties file
```
#!/bin/sh

# clones everything ksql and kstreams
ORIG_TUTORIAL=filtering

# to clone just the ksql part
#ORIG_TUTORIAL=filtering/ksql

# to clone just the kstreams portion
#ORIG_TUTORIAL=filtering/kstreams
NEW_TUTORIAL=something-close-to-filtering
SEMAPHORE_TEST_NAME="My New Filtering Tutorial"
PERMALINK=filtering-new-hotness

### CODE REVIEW

File: templates/ksql/filtered/code/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+      - broker
+    ports:
+      - "8081:8081"
+    environment:
+      SCHEMA_REGISTRY_HOST_NAME: schema-registry
+      SCHEMA_REGISTRY_KAFKASTORE_CONNECTION_URL: 'zookeeper:2181'
+
+  ksql-server:
+    image: confluentinc/cp-ksql-server:<CP-VERSION>
+    hostname: ksql-server
```
Comment (colinhicks) [inline templates/ksql/filtered/code/docker-compose.yml]: I presume the plan would be to tweak this for ksqlDB and ask for a KSQLDB_VERSION?

Comment (bbejeck) [reply templates/ksql/filtered/code/docker-compose.yml]: Yes, I missed that one.

**Comment by bbejeck** [comment]
@colinhicks updated and rebased on master

**Comment by colinhicks** [comment]
Thanks @bbejeck, this looks fantastic. Looking forward to kicking the tires.

**Comment by bbejeck** [comment]
When building a tutorial with `./gen_project.sh` CD into `_includes/<TUTORIAL-NAME>/kstreams/code`  then run

1. `gradle wrapper`
2. `./gradlew build`
3. `/.gradlew idea`
 Then open the project in IntelliJ.  More to come..

File: templates/ksql/filtered/code/docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    ports:
+      - "29092:29092"
+    environment:
+      KAFKA_BROKER_ID: 1
+      KAFKA_ZOOKEEPER_CONNECT: 'zookeeper:2181'
+      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT
+      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://broker:9092,PLAINTEXT_HOST://localhost:29092
+      KAFKA_METRIC_REPORTERS: io.confluent.metrics.reporter.ConfluentMetricsReporter
+      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
+      KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS: 0
```
Comment (colinhicks) [inline templates/ksql/filtered/code/docker-compose.yml]:
```suggestion
      KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS: 0
      KAFKA_TRANSACTION_STATE_LOG_MIN_ISR: 1
      KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR: 1
```

File: templates/ksql/static/dev/code/start-cli.sh [hunk 3]
```diff
@@ -0,0 +1 @@
+docker exec -it ksql-cli ksql http://ksql-server:8088
```
Comment (colinhicks) [inline templates/ksql/static/dev/code/start-cli.sh]:
```suggestion
docker exec -it ksqldb-cli ksql http://ksql-server:8088
```

**Comment by bbejeck** [comment]
Now running `gen_project.sh` with `kstreams` produces a fully running bare-bones tutorial, next is `ksql`

**Comment by bbejeck** [comment]
Updated with:
1. Running `ksql` template project with generation
2. Prints a checklist at the end of the generation or cloning process plus prints a copy of the checklist in the new tutorial directory
3. Moved generation and cloning scripts to a `tools` directory

**Comment by bbejeck** [comment]
@colinhicks after you review the `README` I think we have something we can merge.

File: README.md [hunk 4]
```diff
@@ -78,6 +78,111 @@ Kafka Tutorials is a bit unique in that each tutorial is self-testing. That is,
...
+  1. `gen_project.sh` 
+  2. `clone_tutorial.sh`
+
+Let's describe each script.
+
+### The `gen_project.sh` script
+
+As the name implies, the `gen_project.sh` script generates the minimal structure for a viable Kafka Tutorials tutorial.  The code and text contained in the tutorial are place holders, and you'll update those with your code and writing to complete the tutorial.
+
+The script expects you to supply a properties file for setting the name and other parts of the tutorial.  Heres' an example properties file you can use.
```
Comment (colinhicks) [inline README.md]:
```suggestion
The script expects you to supply a properties file for setting the name and other parts of the tutorial.  Here's an example properties file you can use.
```

File: README.md [hunk 5]
```diff
@@ -78,6 +78,111 @@ Kafka Tutorials is a bit unique in that each tutorial is self-testing. That is,
...
+NEW_TUTORIAL=something-close-to-filtering
+SEMAPHORE_TEST_NAME="My New Filtering Tutorial"
+PERMALINK=filtering-new-hotness-tutorial
+```
+
+To clone a tutorial:
+
+1. Create a branch
+2. Make sure you are in the `Kafka Tutorials` base directory
+3. Execute `> tools/clone_tutoral.sh ~/clone-tutorial-props.sh`
```
Comment (colinhicks) [inline README.md]:
```suggestion
3. Execute `> tools/clone_tutorial.sh ~/clone-tutorial-props.sh`
```

**Pull request merged**